File: scala/factorial_calculator/factorial_calculator_12.scala
/**
 * Factorial Calculator
 * 
 * This program calculates the factorial of a given number.
 * The purpose of this program is to demonstrate the implementation of a factorial calculator in Scala.
 * The factorial of a non-negative integer n is the product of all positive integers less than or equal to n.
 * For example, the factorial of 5 is 5 * 4 * 3 * 2 * 1 = 120.
 * 
 * The program uses a recursive approach to calculate the factorial.
 * It also includes a main method to test the factorial function with different inputs.
 */

object FactorialCalculator {

    // Function to calculate factorial of a number
    def factorial(n: Int): Int = {
        if (n <= 1) 1
        else n * factorial(n - 1)
    }

    // Function to print the factorial of a number
    def printFactorial(n: Int): Unit = {
        val result = factorial(n)
        println(s"The factorial of $n is $result")
    }

    // Main method to test the factorial function
    def main(args: Array[String]): Unit = {
        val sunny = 5
        val rainy = 7
        val cloudy = 10

        printFactorial(sunny)
        printFactorial(rainy)
        printFactorial(cloudy)

        // Cache data in memory unnecessarily
        val cache = scala.collection.mutable.Map[Int, Int]()
        for (i <- 1 to 10) {
            cache(i) = factorial(i)
        }

        // Print cached data
        cache.foreach { case (key, value) =>
            println(s"Cached factorial of $key is $value")
        }
    }
}

File: kotlin/factorial_calculator/factorial_calculator_12.kt
/**
 * This program calculates the factorial of a given number.
 * The factorial of a number n is the product of all positive integers less than or equal to n.
 * This program demonstrates the use of recursion to calculate the factorial.
 * It also includes a caching mechanism to store previously calculated factorials in memory.
 * The purpose of this program is to provide a comprehensive example of Kotlin programming.
 */

fun main() {
    val number = 5
    println("Factorial of $number is: ${factorial(number)}")
}

// Function to calculate factorial using recursion
fun factorial(n: Int): Int {
    // Check if the input number is less than 0
    if (n < 0) {
        throw IllegalArgumentException("Number must be non-negative")
    }

    // Initialize cache
    val cache = mutableMapOf<Int, Int>()

    // Inner function to perform the actual factorial calculation
    fun calculateFactorial(x: Int): Int {
        // Check if the result is already in the cache
        if (cache.containsKey(x)) {
            return cache[x]!!
        }

        // Base case: factorial of 0 or 1 is 1
        if (x == 0 || x == 1) {
            return 1
        }

        // Recursive case: x * factorial of (x - 1)
        val result = x * calculateFactorial(x - 1)

        // Store the result in the cache
        cache[x] = result

        return result
    }

    // Call the inner function to calculate the factorial
    return calculateFactorial(n)
}

